charms/mlflow-server/tests/integration/test_charm.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from pathlib import Path
+from time import sleep
+
+import pytest
+import requests
+import yaml
+from lightkube.core.client import Client
+from lightkube.models.rbac_v1 import PolicyRule
+from lightkube.resources.rbac_authorization_v1 import Role
+from pytest_operator.plugin import OpsTest
+
+log = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+CHARM_NAME = METADATA["name"]
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest):
+    db = "mlflow-db"
+    obj_storage = "minio"
+    await ops_test.model.deploy("charmed-osm-mariadb-k8s", application_name=db)
+    await ops_test.model.deploy(obj_storage)
+
+    my_charm = await ops_test.build_charm(".")
+    image_path = METADATA["resources"]["oci-image"]["upstream-source"]
+    resources = {"oci-image": image_path}
+    await ops_test.model.deploy(my_charm, resources=resources)
+    await ops_test.model.add_relation(CHARM_NAME, obj_storage)
+    await ops_test.model.add_relation(CHARM_NAME, db)
+    await ops_test.model.wait_for_idle(
+        status="active", raise_on_blocked=False, raise_on_error=False
+    )
+
+    assert ops_test.model.applications[CHARM_NAME].units[0].workload_status == "active"
+
+
+async def test_access_dashboard(ops_test: OpsTest, request):
```

Comment:
We can split this test up a bit more to make it clear what role everything plays.  2/3 of the code is really a fixture (set up istio ingress gateway).  At a minimum it would be helpful to denote this in comments, but I think we can address a few things at once here.  

This test checks more than just dashboard access, it checks dashboard access **through an ingress**.  We should test these two things separately (1: is dashboard accessible, 2: is dashboard accessible through the ingress).  That way if say istio breaks something we see the mlflow dashboard is ok but mlflow dashboard via ingress is broken, and vise versa if the dashboard actually breaks we don't debug istio

Reply:
Another way that might be clearer would be to have two fixtures, and pass them in as a parameter:

```
def istio_ingress():
  # deploy ingress
  yield dashboard_url
  # clean up ingress

def null_ingress():
  # does not deploy an ingress
  return dashboard_url

@pytest.mark.parametrize(
  "ingress",
  (
    (pytest.lazy_fixture('istio_ingress'),),
    (pytest.lazy_fixture('null_ingress'),),
  )
)
def test_access_dashboard(ingress):
  dashboard_url = ingress
  ...
```

This uses an extra dependency (pytest-lazy-fixture), but to me reads a bit clearer on what it is doing.  Curious to hear your thoughts!